src/python/pants/backend/experimental/python/lockfile.py

```diff
@@ -80,13 +88,28 @@ def from_tool(cls, subsystem: PythonToolRequirementsBase) -> PythonLockfileReque
             description=f"Generate lockfile for {subsystem.options_scope}",
         )
 
+    @property
+    def hex_digest(self) -> str:
+        """Produces a hex digest of this lockfile's inputs, which should uniquely specify the
+        resolution of this lockfile request.
+
+        Inputs are definted as requirements and interpreter constraints.
+        """
+        m = hashlib.sha256()
+        for requirement in self.requirements:
+            m.update(requirement.encode("utf-8"))
+        for constraint in self.interpreter_constraints:
+            m.update(str(constraint).encode("utf-8"))
+        return m.hexdigest()
```

Comment:
Thinking out loud here..my first thought was "This seems inefficient to iterate over every element! You could hash the whole collection to avoid that".

But I suspect this is intentional to iterate over every element, and that it makes the algorithm more robust to hash collisions? Is that right? If so, maybe add a comment explaining this nuance.

Reply:
Secure hashes from `hashlib` work on `bytes` objects -- you can't hash an iterable, but you can hash a `bytes` representation of every item of an iterable. I don't think there's any nuance here?

Reply:
Hm, @stuhood thoughts?

Reply:
Returning to this, under the hood, `m.update(a + b)` is equivalent to `m.update(a); m.update(b)` (see https://docs.python.org/3/library/hashlib.html) so I presume that `hashlib` is doing something with `stringio` or similar under the hood to build up the input to the digest function, which is about as efficient as what we'd end up doing.

Reply:
To clarify, I'm wondering if we could do `m.update(repr(self.requiremenets).encode("utf-8"))`, so that would not quite be the same thing. I'm pretty sure that's more likely to have a hash collision, only not sure it matters.

Reply:
Using the `repr` of the collection would be equivalent to adding a bit of salt to the hashes; I don't think either approach meaningfully changes the probability of a collision. The chance of a collision over a 256-bit cryptographic hash is vanishingly small either way, so the presence/otherwise of a few delimiters isn't going to meaningfully change that.

Reply:
One thing to consider in this is sort order. Are the requirements/constraints sorted when going into this, otherwise a change of order would yield different hashes, which I suppose is undesirable, as they would still resolve to the same thing once used, iiuc.